<a href="https://colab.research.google.com/github/miraqulass/MachineLearning/blob/MLPractices/RNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RECURRENT NEURAL NETWORKS**

**Natural Language Processing** -- A discipline in computing that deals with communication between natural(human) languages and computer languages.

**Recurrent Neural Network** -- The kind of Neural Network that processes sequential data such as texts or characters.

In [ ]:
vocab = {}
word_encoding = 1

def bag_of_words(text):
  global word_encoding

  words = text.lower().split(" ")
  bag = {}

  for word in words:
    if word in vocab:
      encoding = vocab[word]
    else:
      vocab[word] = word_encoding
      encoding = word_encoding
      word_encoding += 1

    if encoding in bag:
      bag[encoding] += 1
    else:
      bag[encoding] = 1

  return bag

text = "Today I have done something constructive, but I must be consistent!"
bag = bag_of_words(text)
print(bag)
print(vocab)

{1: 1, 2: 2, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1}
{'today': 1, 'i': 2, 'have': 3, 'done': 4, 'something': 5, 'constructive,': 6, 'but': 7, 'must': 8, 'be': 9, 'consistent!': 10}


In [ ]:
voc = {}
word_encoding = 1

def bag_of_wordss(text):
  global word_encoding

  words = text.lower().split(" ")
  bag = {}

  for word in words:
    if word in voc:
      encoding = voc[word]
    else:
      voc[word] = word_encoding
      encoding = word_encoding
      word_encoding += 1

    if encoding in bag:
      bag[encoding] += 1
    else:
      bag[encoding] = 1

  return bag

text = "The broad-based or bread-based government in Kenya is bullshit, should be disbanded! And dissolve the parliament too!"
bag = bag_of_wordss(text)

print(bag)
print(voc)

{1: 2, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1, 11: 1, 12: 1, 13: 1, 14: 1, 15: 1, 16: 1}
{'the': 1, 'broad-based': 2, 'or': 3, 'bread-based': 4, 'government': 5, 'in': 6, 'kenya': 7, 'is': 8, 'bullshit,': 9, 'should': 10, 'be': 11, 'disbanded!': 12, 'and': 13, 'dissolve': 14, 'parliament': 15, 'too!': 16}


In [ ]:
from keras.datasets import imdb
from keras.preprocessing import sequence
import tensorflow as tf
import os
import numpy as np


VOCAB_SIZE = 88584

MAXLEN = 250
BATCH_SIZE = 64

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words = VOCAB_SIZE)

In [ ]:
train_data[0]

[1,
 14,
 22,
 16,
 43,
 530,
 973,
 1622,
 1385,
 65,
 458,
 4468,
 66,
 3941,
 4,
 173,
 36,
 256,
 5,
 25,
 100,
 43,
 838,
 112,
 50,
 670,
 22665,
 9,
 35,
 480,
 284,
 5,
 150,
 4,
 172,
 112,
 167,
 21631,
 336,
 385,
 39,
 4,
 172,
 4536,
 1111,
 17,
 546,
 38,
 13,
 447,
 4,
 192,
 50,
 16,
 6,
 147,
 2025,
 19,
 14,
 22,
 4,
 1920,
 4613,
 469,
 4,
 22,
 71,
 87,
 12,
 16,
 43,
 530,
 38,
 76,
 15,
 13,
 1247,
 4,
 22,
 17,
 515,
 17,
 12,
 16,
 626,
 18,
 19193,
 5,
 62,
 386,
 12,
 8,
 316,
 8,
 106,
 5,
 4,
 2223,
 5244,
 16,
 480,
 66,
 3785,
 33,
 4,
 130,
 12,
 16,
 38,
 619,
 5,
 25,
 124,
 51,
 36,
 135,
 48,
 25,
 1415,
 33,
 6,
 22,
 12,
 215,
 28,
 77,
 52,
 5,
 14,
 407,
 16,
 82,
 10311,
 8,
 4,
 107,
 117,
 5952,
 15,
 256,
 4,
 31050,
 7,
 3766,
 5,
 723,
 36,
 71,
 43,
 530,
 476,
 26,
 400,
 317,
 46,
 7,
 4,
 12118,
 1029,
 13,
 104,
 88,
 4,
 381,
 15,
 297,
 98,
 32,
 2071,
 56,
 26,
 141,
 6,
 194,
 7486,
 18,
 4,
 226,
 22,
 21,
 134,
 476,
 26,
 480,
 5

In [ ]:
print(max([max(sequence) for sequence in train_data]))

88583


In [ ]:
len(train_data[0])

218

In [ ]:
train_data = sequence.pad_sequences(train_data, MAXLEN)
test_data = sequence.pad_sequences(test_data, MAXLEN)

In [ ]:
len(train_data[0])

250

In [ ]:
# Creating the Model

model = tf.keras.Sequential([
    tf.keras.layers.Embedding(VOCAB_SIZE, 32, input_length=MAXLEN),
    tf.keras.layers.LSTM(32),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [ ]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm_1 (LSTM)                        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [ ]:
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=['acc'])

history = model.fit(train_data, train_labels, epochs=10, validation_split=0.2)

Epoch 1/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 9s 12ms/step - acc: 0.7012 - loss: 0.5485 - val_acc: 0.8454 - val_loss: 0.3584
Epoch 2/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 8s 12ms/step - acc: 0.9246 - loss: 0.2005 - val_acc: 0.8644 - val_loss: 0.3334
Epoch 3/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 12s 16ms/step - acc: 0.9636 - loss: 0.1084 - val_acc: 0.8538 - val_loss: 0.3870
Epoch 4/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 8s 13ms/step - acc: 0.9823 - loss: 0.0595 - val_acc: 0.8570 - val_loss: 0.4364
Epoch 5/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - acc: 0.9823 - loss: 0.0553 - val_acc: 0.8642 - val_loss: 0.4958
Epoch 6/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - acc: 0.9875 - loss: 0.0415 - val_acc: 0.8578 - val_loss: 0.7467
Epoch 7/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - acc: 0.9917 - loss: 0.0283 - val_acc: 0.8620 - val_loss: 0.6327
Epoch 8/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 8s 13ms/step - acc: 0.9935 - loss: 0.0245 - val_acc: 0.8522 - val_loss: 0.6619
Epoch 9/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 7s 1

In [29]:
results = model.evaluate(test_data, test_labels)
print(results)

782/782 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - acc: 0.8439 - loss: 0.7652
[0.7687898278236389, 0.8403599858283997]


In [38]:
# Make predictions

from tensorflow.keras.preprocessing.text import text_to_word_sequence
from tensorflow.keras.preprocessing import sequence

word_index = imdb.get_word_index()

def encode_text(text):
  tokens = text_to_word_sequence(text)
  tokens = [word_index[word] if word in word_index else 0 for word in tokens]
  return sequence.pad_sequences([tokens], MAXLEN)[0]

text = "that was a splendid game, so splendid"
encoded = encode_text(text)
print(encoded)

[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0 

In [39]:
# Decoding

reverse_word_index = {value: key for (key, value) in word_index.items()}

def decode_integers(integers):
  PAD = 0
  text = ""

  for num in integers:
    if num != PAD:
      text += reverse_word_index[num] + " "

  return text[:-1]

print(decode_integers(encoded))

that was a splendid game so splendid


In [43]:
def predict(text):
  encoded_text = encode_text(text)
  pred = np.zeros((1, 250))
  pred[0] = encoded_text
  result = model.predict(pred)

  print(result[0])

positive_review = "Action-packed fantasy with a strong lead performance and impressive fight scenes. That was an awesome show, really loved it!"
predict(positive_review)

negative_review = "That movie sucked. I hated it and wouldn't watch it again. Was one of the worst things I've ever watched!"
predict(negative_review)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
[0.9993363]
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
[0.5750962]


**N.B**

*   The model developed has some issues (overfitiing) due to less data used to create the model. Working on it though!
*   The above results classifies the first review (0.9993363) as positive and (0.5750962) as negative. This is due to the problems we're experiencing with the model.

